WIP: Backbone of a tutorial

Bibobu opened the pull request:
**Summary**

This is an initial PR to provide a basic outline for the tutorials and a first step on organizing the project. As Slack channels have finite lifetime, it can also be used as a place to make long term todo lists discuss goals.

**Details**

This PR provides some initial files for discussion in the project of making detailed LAMMPS tutorials for newcomers. In previous discussion on the Slack channel, many different views on the directions the project could take emerged.

Given the variety of profiles of people asking reference, I suggest an initial division in three stages sorted by topic complexity from the very basic to more advanced stuff. If the initial stages shall be guided and pedagogical (maybe in a linear fashion), the latter might be thought as independent sections for more autonomous users. The idea is to start with guided hands-on to autonomous hands-on tutorials, keeping a "do it yourself" approach at the center.

The initial proposed stages are:
* Beginner: Very basic stuff from how-to format input to how-to analyze a log file and visualize a trajectory (ffmpeg? very short intro to OVITO? or VMD?) with variables, compute and fixes introduction in between.
* Advanced: Introduction to advanced stuff from extended hamiltonians and fixes combination to replica exchange or other simulation techniques. Many sections of the how-to could be detailed here to illustrate more general concepts.
* Confirmed: maybe detailed examples of Python coupling or step by step guide of a small library example. This might be an introduction to the developer guide section and a complement to the "coupling LAMMPS to other code" section of the how-to.

This is a work-in-progress as many things need to be settled down, some important axis need to be discussed and most of the proposed sections need to actually be written.

I don't know how this could make use of the Github Project system.

Bibobu committed:
A first commit with some initial description of the
tutorial goals and a first incomplete class on file formats.

Bibobu committed:
Some edits to continue the initial lesson.

Bibobu committed:
Added a basic structure of the tutorial in three steps. Added simple
files for the outline of the tutorial. Added empty files for the first
sections of the Beginner section.

simongravelle reviewed (commented):
Hello,

This is my first time commenting on a pull-request from vscode, so I hope that it will be simple for you to see/accept/reject my comments/suggestions. 

Thank you very much @Bibobu for starting the project with a very nice simple tutorial. I did some tiny eddit to the text, feel free to reject them.

Some very generic comments:
- we should choose between american and british english. I suppose american is the most obvious choice here?
- I would impose from the start the LAMMPS version people have to use to follow the tutorial without trouble. Ideally the last stable version. 
- is '.lmp' a convensional extension for LAMMPS data file? I am personaly using '.data' as it gets recognized directly by MDAnalysis...
- side note: I wonder if using LJ units for the very first tutorial makes it easier for beginners, or harder... on the one hand, we don't have to deal with units conversion, which is great , on the other hand, LJ units are counter intuitive and writting "pair_coeff 1 1 1 1" may seem bizare. I have the same interogation on lammpstutorials.github.io were I also use LJ units on the first tutorial.
 
Best,
Simon

edit: OK it seems that I have pushed my changes directly into the PR. Sorry @Bibobu, that was not my goal, I was hopping that the changes would appear more as "suggestions", as it does on Gitlab.

simongravelle committed:
made some minor changes to the text

Bibobu commented:
Hi @simongravelle,

To answer your questions/remarks:

* I agree american english makes more sense. I was taught british english but tend to mix the two if I am not careful. I'll try to stick to AE and set my spellcheck to do so.
* Asking to use at least the last stable version is a good idea. I think this is the one deployed on major Linux repositories. But I am unsure and I do not know who do the several packaging. This would make things consistent for people downloading and installing it through package managers.
* I personally use `*.lmp` as an extension as this is one of the extension for the vim syntax highlighter provided in the source repo along with `in.*` and `*.in`. I think this is also a convention to settle. This is not very important for LAMMPS itself but it should be consistent throughout the tutorial.
* I have the same question with regards to LJ units. But as a physicist by training I might be biased as dimensionless units feel very natural to me and more convenient. Maybe it is better to start using a real life example and show how to convert from there. I don't know.

I think your changes corrected typos and homogenized the overall text. So it's all good. Besides there is not much really fixed for now.

Bibobu commented:
I also thought about making a "Common problems" files in the same spirit as the one that is already in the manual and the pinned post of the forum but:
* This might be redundant.
* The two mentioned sources might be edited and completed.

However, many people still come to the forum in a "see this, what do?" spirit, so maybe giving a first list of things to check when getting an error in a tutorial might make sense. If this tutorial ends up being linked to LAMMPS main communication channels this might be the place where people head to before the forum or the manual. (Not getting optimistic here, simply pragmatic.)

akohlmey commented:
@Bibobu thanks for moving this forward.
I am a bit biased in this respect, but I think for the Beginners section it may be advantageous to use the LAMMPS-GUI first.
It wasn't ready for the LAMMPS workshop in August, but I've been receiving good feedback from multiple people. This can't do what the specialized tools can do, but is sufficient for the simple stuff if used in a smart way. Thus learning how to use OVITO or VMD or other tools for system generation, analysis, and visualization could be done toward the end of the section. If needed, I can add missing functionality, but at the moment there should not be much left except for features that are very difficult to add.

A "Common problems and solutions" document can be hosted anywhere and then other places that people might find would contain a link to that. I would not worry too much about people not search for this first. Some people cannot be helped, they will always post immediately if they encounter an issue. But everybody else that has not looked in the right place or not looked at all could be embarrassed by being pointed to the page and then look at it in the future (provided it contains the information being looked for).

Bibobu committed:
Added labels to files at the top.

Bibobu committed:
Changes visualizing file name to american english spelling

Bibobu committed:
Corrected a file name in Beginner.rst

Bibobu committed:
Adding the end of the file setting_up_simulations and reference to
running_and_post_processing.

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+****************
+
+The first file of the two is usually referred to as a `data file`. Its format
+is rather strict and you can see that there are general information at the
+top of the file and several sections starting with a capitalized word, a blank
+line and some numbers. Let's slowly go through all of this.
```
```suggestion
is rather strict, containing:
* a "header" block, which must come first in the file, and
* several "data sections", each of which:
  * starts with a capitalized keyword (such as "Atoms" or "Velocities")
  * followed by a blank line
  * and then a block of numbers (such as each atom's ID, x, y, and z velocities in "Velocities")

Let's slowly go through all of this.
```

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+the number of `atoms` (125), the number of `atom types` (1) and three lines
+containing `xlo xhi` like indications. This header is simple, but generally,
+headers can contain much more information. The first two lines are explicit,
+you define a system with 125 atoms all of which have the same characteristics.
+The last three lines define the volume in which these atoms are contained. It
```
```suggestion
you define a system with 125 atoms all of which have the same characteristics.
They all have the same characteristics because your simulation has only one `atom type`;
you will soon learn how to specify different interactions between particles
in LAMMPS by giving the particles different atom types.
The last three lines define the volume in which these atoms are contained. It
```

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+   Section input # Number of line and format depend on the section.
+
+The first section you should see is the `Masses` section. In LAMMPS, masses are
+assigned to atom types so you only have one line here. All types have their own
+masses but several types can have the same mass. Types are LAMMPS way to refer
+to properties of particles that are *the same* for all particles of *the same
+type*. They are also used to determine how particles interact with one another.
+Types are not always binded to chemical species and you will see in further
+tutorials how it can be convenient to define different types for similar atoms.
```
```suggestion
usually assigned by atom types. Since you have only one atom type
in this simulation, this `Masses` section has only one line.
```

"usually" added because LAMMPS can have per-atom `rmass` (unlike GROMACS for example). Types are introduced earlier, and detailed information about them can be pushed to when we actually introduce multitype simulations.

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+In an empty directory, open a text file and copy the following text inside:
+
+.. code-block:: LAMMPS
+
```
```suggestion

atom_style atomic
```

My preference is, since we introduce `atom_style` a bit later, we should make this line explicit. Alternatively we could delay discussing `atom_style` at all until (for example) we get to a molecular simulation.

Bibobu replied:
My problem with delaying the discussion about this command would be that the `# atomic` style will be mentioned in the data file we go through later on. I think you are right in making it explicit, as it will make people notice this settings exist. We can mention that a detailed discussion about the several atom types will be given later in the tutorial.

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+the particles along the x, y and z axis in that order. The particle ID refers
+to the same ID as in the `Atoms` section.
+
+The format and meaning of the number in the `Atoms` and `Velocities` sections
+will depend on the `atom_style` the code is told to use. In most recents
+versions of LAMMPS, writing a data file with `write_data` will append a comment
+next to the `Atoms` section name (here `# atomic`). This helps human readers to
+know the meaning of the different number. Some formats can include more
+information like a molecule tag, particles' charges, particles' spins or
+orientation etc. For the velocities, some `atom_style` can require keeping
+track of angular momentum, angular velocities etc. You'll find a detailed
+description of each format in the `read_data section of the manual`_.
```
```suggestion
This data file is quite simple because our Very First LAMMPS Script had the
`atom_style` `atomic`, where each particle only has an ID number 
and a three-dimensional position and velocity (and periodic image location).
More detailed simulations will use other `atom_style` commands,
such as `atom_style molecular`, which also tracks molecule ID
numbers and intramolecular arrangements like bonds,
angles, "dihedrals" and "impropers" (which you will learn about later).
You can see that LAMMPS includes a comment `# atomic` next to the
`Atoms` section name (here `# atomic`), and you can find a detailed
description of each format in the `read_data section of the manual`_.
```

We regularly instruct our forum readers to use the latest LAMMPS version and we should assume they are doing the same when reading this manual (especially if they are using LAMMPS GUI).

Bibobu replied:
This is a very good remark. I will also add it in the introduction of the tutorial where I also tell readers to get in touch with their IT support to get LAMMPS installed. I think it might be necessary to mention the latest release is preferable.

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+orientation etc. For the velocities, some `atom_style` can require keeping
+track of angular momentum, angular velocities etc. You'll find a detailed
+description of each format in the `read_data section of the manual`_.
+
+As a first takeaway, remember that `data files` contain detailed information
+on a simulation system at a given time. They are more convenient for input and
+output. Several simulation softwares allow you to export files as LAMMPS
+`data files` and take them as input. But data format are not straightforward to
+use for analyses: they are heavy and may contain useless information. This is
+what the dump file format is for.
```
```suggestion

You may have noticed our first LAMMPS script also produced a `dump file`.
We will now read through this dump file, and then learn how it is
different from a data file and how to use each file type.
```

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+  5 1 0.8 0 0
+  ...
+
+The format is more simple compared to the data file. Each section is labeled
```
```suggestion
The dump file format is simpler than the data file. Each section is labeled
```

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+with a header directly followed by the data we wanted to dump. Here we used the
+basic atom dump_style so we only have atoms' id, types and scaled coordinates
+(that is coordinates divided by box length in each dimension).
+
```
```suggestion

From the TIMESTEP heading, you might guess that a dump file will
usually contain information about how a simulation changes `over time`,
and you would be correct! Thus, a `data` file is used to store the `complete`
state of a simulation with the `write_data` command, and in later lessons
you will see how to start a new simulation from that state with the
`read_data` command. On the other hand, a `dump` file stores information
about how the system changes over time, and can then be used for analyzing
simulation results.

If you (or your supervisor) have previously used other molecular dynamics
software, you may recognize that the dump file loosely corresponds to "coordinates"
or "trajectory" output files from other software. However, LAMMPS gives you
great flexibility in what you choose to output. For example, the default `dump` format
outputs scaled coordinates, but we will soon see how to output unscaled
coordinates instead. You may know that other molecular dynamics packages
store information that `does not` change throughout a simulation (such as
molecular bonds and particle charges) in "parameter", "topology" or
"configuration" files. In LAMMPS, this information is read from a `data` file, but
the `data` file also usually contains coordinates and velocities.
```

Bibobu replied:
Great addition.

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+basic atom dump_style so we only have atoms' id, types and scaled coordinates
+(that is coordinates divided by box length in each dimension).
+
+You can compare the `dump file` with the `data file`, and
+see that they basically contain the same information, with few
+exceptions, namely the total number of types, masses and the velocities. Getting only the
+scaled coordinates of the atoms is not alway ideal, but fortunately we can get
+much more.
```
```suggestion
```

srtee commented on src/file_format.rst:
```diff
@@ -0,0 +1,189 @@
+scaled coordinates of the atoms is not alway ideal, but fortunately we can get
+much more.
+
+In your `in.lmp` file, replace the `write_dump` line with the following:
```
```suggestion
As an example of customizing the dump file, 
in your `in.lmp` file, replace the `write_dump` line with the following:
```

srtee commented on src/setting_up_simulations.rst:
```diff
@@ -0,0 +1,128 @@
+4. putting atoms at the coordinates defined by the lattice.
+5. giving them a mass, and writing output files.
+
+That's it. Nowhere are interaction between atoms defined. As such they do not
```
```suggestion
That's it. Nowhere are interactions between atoms defined. As such they do not
```

srtee commented on src/setting_up_simulations.rst:
```diff
@@ -0,0 +1,128 @@
+**after** the number of `atom types` is defined in the simulation.
+
+It is worth detailing the format of the `pair_coeff` command. It defines
+long range interactions between atoms with regard to their types. That is the
```
```suggestion
how nearby atoms interact, based on their types. That is the
```

"Long range" will get people confused with kspace later!

srtee commented on src/setting_up_simulations.rst:
```diff
@@ -0,0 +1,128 @@
+and energy for our system. In the general case, initial geometries are guesses
+from chemical information from *real* systems, but they're not necessary
+representative of equilibrium state of the model with the interactions you
+are using. You can end-up in a situation where local structure cause force
+divergence and numerical integration becomes unstable. Bad initial geometry
+is a common problem that many people encounter without realizing.
```
```suggestion
are using. In a "bad" geometry, "clashes" between particles that are too close to each other
result in unphysically large forces and unstable numerical integration. If
you are new to molecular dynamics and find your simulation crashing, but
you know you have reasonable parameters (for example, they were accurately
copied and converted from other previously published papers), you should
investigate if your initial structure is bad, both by "visualising" the structure
-- that is, exploring a 3D view in a program like VMD or Ovito -- and by
trying an energy minimization, as we now describe.
```

srtee reviewed (commented):
General editing, with the comment that (for me at least) in good teaching, we try to give as _little_ information to the student as possible at any one time, on the assumption that it will already be a lot of information for them to absorb -- some statements seem simple to an expert because we "don't know what we know" but will be very confusing to a student. For example, maybe we should completely remove all `atom_style` references and introduce `atom_style` in a next lesson, for example, when showing them a dimer molecule.

srtee commented:
> @Bibobu thanks for moving this forward. I am a bit biased in this respect, but I think for the Beginners section it may be advantageous to use the LAMMPS-GUI first.

I agree about LAMMPS-GUI. I think it has the right amount of visual guidance and UI "sugar", enough to be really helpful but not so much that it starts interfering.

I had an RA who basically taught herself LAMMPS using @simongravelle 's great [tutorials](https://lammpstutorials.github.io/), running them in LAMMPS-GUI. She did it over a month of calendar time, but in between her other work she only spent about 40 hours - one full-time week - with about four hours (spread out) of help and trouble-shooting from me. At the end I reckon she would've been at PhD starting level -- that is, able to perform and analyze MD runs with me providing scripts and usual supervisory feedback (sadly I didn't have the funding to keep her). Even if she was in the upper decile when it comes to raw smarts (which I don't doubt), she was coming to this with zero prior computational experience, so the utility of LAMMPS-GUI (and Simon's tutorials!) for teaching is very clear.

As an alternative, we should show people _right away_ how to visualize configurations in VMD / Ovito and plot thermo data over time in Python -- otherwise it's impossible for them to have any idea what they are doing in LAMMPS, or indeed in molecular dynamics in general. Every MatSci newcomer who posts desperately for help but hasn't learned how to visualize and plot is wasting their precious, precious time (and ours too).

akohlmey commented:
> As an alternative, we should show people _right away_ how to visualize configurations in VMD / Ovito and plot thermo data over time in Python -- otherwise it's impossible for them to have any idea what they are doing in LAMMPS, or indeed in molecular dynamics in general. Every MatSci newcomer who posts desperately for help but hasn't learned how to visualize and plot is wasting their precious, precious time (and ours too).

Some thoughts on this:
- if VMD or OVITO are installed you can load the system into them right away with "Run->View in VMD" or "Run->View in OVITO" (or CTRL-SHIFT-D and CTRL-SHIFT-O, respectively)
- plotting of thermo data is included and automatic, if the thermo command is set to a reasonable value
- one can create a "live animation" if there is a "dump image" command. However, the syntax for that can be quite complex. Thus I have been thinking to add a button to the Image viewer (where size, zoom, view angles, group, and some visualization settings can be customized) that would either copy a suitable command line into the cut-n-paste buffer, so it could be added to the input with CTRL-V, or just insert it before the first *run* command with some suitable comments with further instructions and so that it can be also be removed with a button as well.

With this one could get pretty far and the interface to VMD/OVITO would allow for a smooth transition. At first, people only would need to learn how to *add* trajectory data to the single frame system from the data file that is used.
In my personal preference the teaching of how to use "advanced" tools should be either at the end of the introductory part or in an intermediate section where also external pre-post processing tools are introduced.

Bibobu committed:
Added a mention to the general [Classics] ref.

Bibobu committed:
Homogen. references to commands, added some [Classics] ref.

Bibobu committed:
Continued the writing of running_and_pp.

Bibobu committed:
Revert "Continued the writing of running_and_pp."

This reverts commit c35346845b037081055b627858a1422b8f365fb6.

Bibobu committed:
Revert "Homogen. references to commands, added some [Classics] ref."

This reverts commit d8c1bc0aaf080a26211978bc99c56b16b570bf7a.

Bibobu committed:
Revert "Added a mention to the general [Classics] ref."

This reverts commit 8cd58c31e58d5eda7c29de2ebf624939a782fad6.

Bibobu commented:
Thanks for your inputs @srtee. I continued writing and didn't see your comments so I committed and reverted back my changes. I'll take a detailed look at the conversation during the week.

To comment on @akohlmey answer, I think the "visualisation" part which IMO should come after analyzing thermo data should give a large part to the use of LAMMPS-GUI in a dedicated section. But I think it should be introduced after showing how to produce images in LAMMPS through the `dump` command. Also VMD and OVITO also provide a lot of computation options but, in my opinion, it is important to emphasize on time averaging in analyzing MD data, as I think many people tend to get stuck on "instantaneous values" computation due to these software limitation. I think visualization acts as a strong distraction with regard to this, so I would carefully introduce them later. LAMMPS-GUI allows for better understanding of what I consider a proper workflow. Note that this is not to downplay VMD or OVITO which are great, but I think self-teaching of these is prone to bad habits. (Let's not mention people drawing conclusion from instantaneous images without proper analysis of configuration when this is clearly needed.)

Any thought?

akohlmey commented:
@Bibobu I suspect we are talking about two different types of visualization here. What LAMMPS-GUI can do and is good for is a quick look of the kind "does my system look like what I expect?". This is often extremely useful when people create systems from the "create_atoms" command with regions. Also, as I mentioned, I think an addition to the Image viewer in the LAMMPS GUI that provides the command line that looks good (the defaults of dump image are often not good) can be helpful.

Any *serious* visualization can come (much) later. The issue is that a lot of problems with inputs created by beginners would be better understood, if they would look at a rendering of their system. A common thing among beginners seems to be (or at least those that post their questions on MatSci) that they assume the system was created according to their expectations, even if the energies say it wasn't. For example, they didn't notice the need for using "units box" or similar. That is why LAMMPS GUI has the features it has minus what would be nice to have but could not be implemented due to limitations of the library interface or my programming skills or available time and interest.

Bibobu committed:
Update src/file_format.rst

I agree with this suggestion.

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/file_format.rst

I agree with this suggestion.

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/file_format.rst

I agree with this suggestion. It is more coherent overall.

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/file_format.rst

Accepted suggestion. See my comment concerning the alternative of leaving it out all together.

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/file_format.rst

I agree with the rephrasing. See my response.

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/file_format.rst

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/file_format.rst

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/file_format.rst

Given previous accepted changes, it makes sense to also accept this one.

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/file_format.rst

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/setting_up_simulations.rst

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/setting_up_simulations.rst

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu committed:
Update src/file_format.rst

Co-authored-by: Shern Tee <shernren@gmail.com>

Bibobu commented:
@srtee I finally found time to take a look at your suggestion and agreed with them all. I agree that it is complicated to find a balance between what to tell and what not to tell to avoid overloading readers with information. However, I take into account that a written tutorial can be read at one's own pace, contrary to a class from which notes might be less complete. I would like to keep the discussion introducing the `atom_style` command but I agree that it might just conclude with something along the lines  "Keep in mind it exists, more on it later."

Bibobu committed:
Some rst formating in file_format.rst.

Bibobu committed:
Re-added the mention of [Classics] MD texts.

Bibobu committed:
Mentioned the last version of LAMMPS is assumed in the text.

Bibobu committed:
Added links to several commands' manual pages.

Bibobu committed:
Added mention of [Classics].

Bibobu committed:
Continued writing of running and post_processing.

simongravelle commented:
> @srtee I finally found time to take a look at your suggestion and agreed with them all. I agree that it is complicated to find a balance between what to tell and what not to tell to avoid overloading readers with information. However, I take into account that a written tutorial can be read at one's own pace, contrary to a class from which notes might be less complete. I would like to keep the discussion introducing the `atom_style` command but I agree that it might just conclude with something along the lines "Keep in mind it exists, more on it later."

What about colored text block (even possibly drop-down text block)  just like the one used in the LAMMPS documentation for the Note and the Warning ? It could be a class "Info" that give more information to the interested reader, but can also be skipped without affecting the results of the tutorial?

Bibobu commented:
> What about colored text block (even possibly drop-down text block) just like the one used in the LAMMPS documentation for the Note and the Warning ? It could be a class "Info" that give more information to the interested reader, but can also be skipped without affecting the results of the tutorial?

That would meet us halfway. If @srtee is okay with that I can make an edit and add a section in the intro to mention that some "Info" blocks can be skipped on first read. I think a block name like "Advanced topic" would be more explicit to the reader. But it will then be important to not overload the text with those. ^^"

srtee commented:
Hi! Coloured text blocks would work fine, I'm okay with that. Thanks for all the thought and discussion from everyone.
